cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -501,13 +593,11 @@ template <typename Rep1, Radix Rad1>
 CUDA_HOST_DEVICE_CALLABLE fixed_point<Rep1, Rad1> operator+(fixed_point<Rep1, Rad1> const& lhs,
                                                             fixed_point<Rep1, Rad1> const& rhs)
 {
-  auto const rhsv = lhs._scale > rhs._scale ? detail::shift_with_precise_round<Rep1, Rad1>(
-                                                rhs._value, scale_type{lhs._scale - rhs._scale})
-                                            : rhs._value;
-  auto const lhsv = lhs._scale < rhs._scale ? detail::shift_with_precise_round<Rep1, Rad1>(
-                                                lhs._value, scale_type{rhs._scale - lhs._scale})
-                                            : lhs._value;
-  auto const scale = lhs._scale > rhs._scale ? lhs._scale : rhs._scale;
+  // TODO: Use structure bindings when C++17 upgrade happens
+  auto const values =
+    detail::get_same_scale_values<Rep1, Rad1>(lhs._value, rhs._value, lhs._scale, rhs._scale);
+
+  auto const scale = lhs._scale < rhs._scale ? lhs._scale : rhs._scale;
```

Comment:
It might be nice to return the scale from `get_same_scale_values` (as the third element of a tuple?). What do you think?

Reply:
Hm, actually maybe another way to do this would be to add a `rescale` method to `fixed_point`, so you could write something like this:
```cpp
auto const scale = scale_type{std::min(lhs._scale, rhs._scale)};
auto const value = lhs.rescale(scale)._value + rhs.rescale(scale)._value;
return fixed_point<Rep1, Rad1>{scaled_integer<Rep1>{value, scale}};
```

```cpp
fixed_point<Rep, Rad> rescale(scale_type scale) const {
  auto const value = detail::shift_with_precise_round<Rep, Rad>(_value, scale_type{scale - _scale});
  return fixed_point<Rep, Rad>{scaled_integer<Rep>{value, scale}};
}
```

Reply:
lol, only saw the first comment and pushed the changes this morning. I think a public `rescale` method might be nice. I recall seeing a couple DecimalType numbers with rescale methods. Interested to get @harrism and @jrhemstad thoughts before i make the change.

Reply:
I played around with CNL a little bit to figure out how it exposes rescaling and rep values:
```cpp
    // Rescaling
    auto a = cnl::fixed_point<int, -4, 10>{1.2345};
    auto b = cnl::fixed_point<int, -2, 10>{a}; // 1.23

    // Conversion to and from rep
    int a_rep = cnl::to_rep<decltype(a)>{}(a); // 12345
    auto c = cnl::from_rep<cnl::fixed_point<int, -2, 10>, int>{}(a_rep); // 123.45
```
I suppose instead of `fixed_point<Rep, Rad> rescale(scale_type)` we could have a constructor similar to CNL's like `fixed_point(fixed_point, scale_type)`.

The interfaces of `cnl::to_rep` and `cnl::from_rep` seem kind of awkward though. We currently have something similar to `from_rep` in the `fixed_point(scaled_integer<Rep>)` constructor, so maybe a corresponding conversion operator to `scaled_integer<Rep>` would be nice.

Reply:
I'm partial to the `rescale` option. I think it is much more intuitive.

Reply:
_From an offline discussion:_ As it returns a new `fixed_point` and isn't a modifying method, the name of the method should not be a verb. `rescaled` is a better name